## 🎯 Learning Outcomes

### By the end of this module, you will be able to

- Understand the concept of **Python modules and libraries**.
- Differentiate between various **module import methods** (`import module_name`, `import module_name as alias`, `from module_name import item`, `from module_name import *`).
- Explain the **advantages and disadvantages** of different import strategies.
- Utilize **functions and constants** from the built-in `math` module to perform mathematical operations.
- Apply common `math` module functions like `sqrt()`, `pow()`, `ceil()`, `floor()`, `fabs()`, and `factorial()`.

---

Python modules are files containing Python code (functions, classes, variables). They allow you to logically organize your code and reuse it across different programs. A Python library is a collection of modules that offer a set of functionalities without you having to write the code yourself.

One common built-in module is `math`, which provides mathematical functions and constants.

Python provides several ways to import modules:

1.  **`import module_name`**: Imports the entire module. You then access its contents using `module_name.item` (e.g., `math.sqrt()`).
2.  **`import module_name as alias`**: Imports the entire module but assigns it a shorter alias. You access contents using `alias.item` (e.g., `np.array()` for NumPy).
3.  **`from module_name import item1, item2`**: Imports only specific items (functions, classes, variables) directly into the current namespace. You can then use them without the module prefix (e.g., `sqrt()` instead of `math.sqrt()`).
4.  **`from module_name import *`**: Imports all items from the module into the current namespace. This is generally discouraged in larger projects as it can lead to name conflicts.

Let's see examples for each import type using the `math` module.

### 1. `import module_name` (e.g., `import math`)

In [1]:
import math
print(math.sqrt(25))

5.0


### 2. `import module_name as alias` (e.g., `import math as m`)

In [2]:
import math as m
print(f"Using alias 'm' for math: m.pow(2, 3) = {m.pow(2, 3)}")

Using alias 'm' for math: m.pow(2, 3) = 8.0


### 3. `from module_name import item1, item2` (e.g., `from math import sqrt, pi`)

In [3]:
from math import sqrt, pi
print(f"Directly importing sqrt and pi: sqrt(36) = {sqrt(36)}, pi = {pi}")
# Note: You can now use sqrt() and pi directly without the 'math.' prefix

Directly importing sqrt and pi: sqrt(36) = 6.0, pi = 3.141592653589793


### 4. `from module_name import *` (e.g., `from math import *`)

In [4]:
from math import *
print(f"Importing all from math: cos(0) = {cos(0)}, factorial(5) = {factorial(5)}")
# WARNING: This practice is generally discouraged in larger projects
# because it can lead to name collisions and make code harder to read and debug.

Importing all from math: cos(0) = 1.0, factorial(5) = 120


## Notable Functions in the `math` Module

Here's a table summarizing some frequently used functions and constants from the `math` module:

| Method/Constant | Description | Example (Output) |
| :-------------- | :------------------------------------------- | :-------------------- |
| `math.sqrt(x)`  | Returns the square root of `x`.             | `math.sqrt(9)` (3.0)  |
| `math.pow(x, y)`| Returns `x` raised to the power `y`.        | `math.pow(2, 3)` (8.0) |
| `math.pi`       | The mathematical constant π (pi).           | `math.pi` (3.14159...) |
| `math.e`        | The mathematical constant e (Euler's number).| `math.e` (2.71828...) |
| `math.ceil(x)`  | Returns the smallest integer greater than or equal to `x`. | `math.ceil(4.2)` (5) |
| `math.floor(x)` | Returns the largest integer less than or equal to `x`.    | `math.floor(4.9)` (4) |
| `math.fabs(x)`  | Returns the absolute value of `x`.          | `math.fabs(-5)` (5.0) |
| `math.factorial(x)` | Returns the factorial of `x`.               | `math.factorial(5)` (120) |
| `math.log(x, base)` | Returns the logarithm of `x` to the given `base`. | `math.log(100, 10)` (2.0) |
| `math.sin(x)`   | Returns the sine of `x` (x in radians).     | `math.sin(math.pi/2)` (1.0) |

In [5]:
import math

Let's demonstrate some functions from the `math` module.

In [6]:
# Using math.sqrt() to calculate square root
number = 16
sqrt_value = math.sqrt(number)
print(f"The square root of {number} is: {sqrt_value}")

The square root of 16 is: 4.0


In [7]:
# Using math.pi for the value of Pi
radius = 5
circle_area = math.pi * (radius ** 2)
print(f"The area of a circle with radius {radius} is: {circle_area}")

The area of a circle with radius 5 is: 78.53981633974483


In [8]:
# Using math.ceil() and math.floor()
value = 4.7
ceiling_value = math.ceil(value)
floor_value = math.floor(value)
print(f"Original value: {value}")
print(f"Ceiling value (rounds up): {ceiling_value}")
print(f"Floor value (rounds down): {floor_value}")

Original value: 4.7
Ceiling value (rounds up): 5
Floor value (rounds down): 4


This demonstrates how to import a module and use its functions and constants to perform common mathematical operations.

---

## Summary

| Concept | Key Point |
| :------------------- | :------------------------------------------------------------------------------------------------------------- |
| **Python Modules** | Files containing Python code (functions, classes, variables) for code organization and reuse. |
| **`import module_name`** | Imports the entire module; access items using `module_name.item`. |
| **`import module_name as alias`** | Imports the entire module with a shorter alias; access items using `alias.item`. |
| **`from module_name import item`** | Imports specific items directly into the current namespace; use items without the module prefix. |
| **`from module_name import *`** | Imports all items from a module; generally discouraged due to potential name conflicts. |
| **`math` Module** | A built-in module providing mathematical functions (`sqrt`, `pow`, `ceil`, `floor`, `fabs`, `factorial`, `log`, `sin`) and constants (`pi`, `e`). |

### Contributed by: Abdulhadi Zubailah